proposed/2022/vulnerabilities-in-restore.md

```diff
@@ -0,0 +1,391 @@
+This is the primary motivation for having 1 error message per package.
+
+### Technical explanation
+
+During every real restore, i.e. one that regenerates the assets file, NuGet will check all sources that provide vulnerability information and raise warnings for the packages that do.
+
+*Which sources can provide vulnerability information?*
+Only V3 http sources can provide vulnerability information. No additional means of providing vulnerability information are proposed here.
+
+#### Technical background
+
+To best understand the technical proposal, it's important to understand the [NuGet package installation process](https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process).
+
+When a package is seen, the client talks to sources and downloads whichever versions are needed. When talking to V3 sources, the client use the `package content` [resource](https://docs.microsoft.com/en-us/nuget/api/package-base-address-resource), which is a static resource that only has information about the available package versions. Only what is contained *within* the package itself affect restore. Nothing else. This allows an equivalence between `http` and `local` sources.
+
+#### Dedicated vulnerability information resource
+
+Checking vulnerability will always take a performance hit, but we want to minimize that. We can:
+
+- Make fewer http calls.
+- Minimize the amount of data we are handling.
+
+Both of these can be achieved by adding a means to understand whether a source supports vulnerability. We'd add a resource such as the following:
+
+```json
+    {
+      "@id": "https://nikolchevulntest.blob.core.windows.net/newcontainer/vulnerabilityindex.json",
+      "@type": "VulnerabilityInfo/6.3.0",
+      "comment": "The endpoint for discovering information about vulnerabilities of packages in this package source."
+    },
+```
+
+When NuGet restores, we check whether a source supports vulnerabilities and if so it can go through a protocol of getting this information.
+
+The client needs all the data to make decisions about vulnerability. In order to not have the client redownload a file and limit the total number of HTTP calls we will ever make, the vulnerability resource will be an index that can contain multiple pages.
+
+- The vulnernability resource **must** be an array of objects.
+- Each object **must** contain:
+  - `@name`, a user friendly name for the page, **must** be unique.
+  - `@id`, the url that contains the data.
+  - `@updated`, a UTC timestamp when the content at `@id` was updated last.
+  - `comment`, a user friendly description.
+- The vulnerability resource **must** at minimum contain 1 page.
+- The pages within the vulnerability resource **must** be exclusive.
+- The vulnerability resource may 2 pages.
+  - One page, `base`, represents the data up to a certain point in time. This page **must** be updated periodically, based on a trigger.

```

Comment:
How big will nuget.org's `base` page be? How quickly will it grow? In the worst case scenario (like an Australian customer), are we concerned about the performance impact of updating this `base` page during a restore?

Reply:
These were the numbers a few weeks ago:

vulnerability_all_alternate.json    2.4 Mb
vulnerability_all_alternate.json.gz 78.9 Kb

> are we concerned about the performance impact of updating this base page during a restore?

Definitely a consideration. We'd only redownload when restore actually happens. The whole process takes time, but right now we think those numbers are relatively small and a cost worth paying given the benefit.

Reply:
FWIW, I tested a different file format for the vulnerability data, trading off computation speed vs file size, with the idea that smaller file size will parse more quickly, so the slower per-package computation might still come ahead. Benchmarks suggest yes, overall perf is better, and the current file size for all github reviewed vulnerabilities is 175KB uncompressed, 9KB gzipped. So, we'll be going ahead with the smaller file size format instead now.